### PR TITLE
[Refactor] Simplify data loading logic around replacing sampler to prevent confusion

### LINKED ISSUE #9655: Sampler is being replaced even with trainer.replace_sampler_ddp=False
## 🐛 Bug

In the following line:

https://github.com/PyTorchLightning/pytorch-lightning/blob/4bebf82f271fb577ed41831403551f0bff97651b/pytorch_lightning/trainer/data_loading.py#L152

PTL tries to replace the sampler object even if `trainer.replace_sampler_ddp` is `False`:

https://github.com/PyTorchLightning/pytorch-lightning/blob/4bebf82f271fb577ed41831403551f0bff97651b/pytorch_lightning/trainer/data_loading.py#L135

### To Reproduce

Should be easy to reproduce with any custom Sampler class that has different init args from the PyTorch default.

### Expected behavior

Function in line 152 should not be called and moved just below 147. This is how it was in version v1.4.8:

https://github.com/PyTorchLightning/pytorch-lightning/blob/1.4.8/pytorch_lightning/trainer/data_loading.py#L143

### Environment

- PyTorch Lightning Version (main):
- PyTorch Version (e.g., 1.10)
- Python version: 3.8.10
- OS (e.g., Linux): Linux
- CUDA/cuDNN version: N/A
- GPU models and configuration: N/A
- How you installed PyTorch (`conda`, `pip`, source): source
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #9655

### Does your PR introduce any breaking changes? If yes, please list them.

This PR remove `trainer.replace_sampler` as the function leaked to do more with samplers, datasets wrapping which might have un-wanted effects on the user side.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/data_loading.py [hunk 1]
```diff
@@ -117,39 +117,121 @@ def auto_add_worker_init_fn(self, dataloader: DataLoader) -> None:
...
+
+    @staticmethod
+    def _prepare_dataloader(dataloader: DataLoader, sampler, mode: Optional[RunningStage] = None) -> DataLoader:
+        dl_kwargs = TrainerDataLoadingMixin._get_dataloader_init_kwargs(dataloader, sampler, mode=mode)
+        dl_cls = type(dataloader)
+        dataloader = dl_cls(**dl_kwargs)
+        return dataloader
+
+    def _determine_sampler(self, dataloader: DataLoader, shuffle: bool, mode: Optional[RunningStage] = None) -> Sampler:
+        if self._requires_distributed_sampler(dataloader):
```
Comment (ananthsub) [inline pytorch_lightning/trainer/data_loading.py:165]: why is this check repeated twice?

**Review by ananthsub — COMMENTED** [review]
@tchaton is there a description of what the dataloader processing pipeline should look like, how it varies for across the different trainer functions (given prediction has a special case here). That could be where some of the confusion around this logic creeps in now.

Is there any expected relationship to the [training type plugin's process_dataloader ](https://github.com/PyTorchLightning/pytorch-lightning/blob/131176b9f5e544a1133240a46dca1960308b7849/pytorch_lightning/plugins/training_type/training_type_plugin.py#L221-L227)? Given the two have similar names, and there's `trainer.prepare_dataloader` and `trainer.training_type_plugin.process_dataloader` one might expect the two to go hand in hand (but they don't)


Another risk with the trainer mixin design is that these are all technically part of the trainer's public API. so removing `auto_add_sampler` is a breaking API change. the odds of someone depending on this outside of the trainer entry functions is low, but how else can we mitigate this? Some ideas:
- Short term: Move all staticmethods out of the mixin into utilities files, and ensure there's testing for them
- Short term: make all methods in the dataloading mixin protected so nothing is publicly exposed going forward

cc @ninginthecloud

File: pytorch_lightning/trainer/data_loading.py [hunk 2]
```diff
@@ -117,39 +117,121 @@ def auto_add_worker_init_fn(self, dataloader: DataLoader) -> None:
...
             return dataloader
 
+        # the DataLoader should be re-created only if we need to inject
+        # the fault tolerant components or the distributed sampler or we need to track indices for predictions.
         if (
-            self.accelerator_connector.replace_sampler_ddp
-            and self.accelerator_connector.is_distributed
-            and not isinstance(dataloader.sampler, DistributedSampler)
-            and not has_iterable_dataset(dataloader)
+            _fault_tolerant_training()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/data_loading.py:148]: did we have this check before? id does not look like it?

**Review by awaelchli — APPROVED** [review]
In a refactor, we are normally not allowed to make changes to tests. Do the tests pass without the modifications/additions?

**Review by carmocca — APPROVED** [review]
LGTM other than open comments 👍

File: pytorch_lightning/trainer/data_loading.py [hunk 3]
```diff
@@ -117,44 +117,69 @@ def auto_add_worker_init_fn(self, dataloader: DataLoader) -> None:
...
-            # use current sampler
-            sampler = dataloader.sampler
+            return self._get_distributed_sampler(
+                dataloader, shuffle, mode=mode, overfit_batches=self.overfit_batches, **self.distributed_sampler_kwargs
+            )
 
-        dataloader = self.replace_sampler(dataloader, sampler, mode=mode)
-
-        return dataloader
+        return dataloader.sampler
```
Comment (ninginthecloud) [inline pytorch_lightning/trainer/data_loading.py:170]: `_determine_sampler` basically helps to generate distributed_sampler, so there's no need to return `dataloader.sampler`. If the dataloader does not require distributed_sampler, we should not update dataloader again.

File: pytorch_lightning/trainer/data_loading.py [hunk 4]
```diff
@@ -298,13 +328,13 @@ def reset_train_dataloader(self, model: Optional["pl.LightningModule"] = None) -
                     "You requested to overfit but enabled training dataloader shuffling."
                     " We are turning off the training dataloader shuffling for you."
                 )
-                self.train_dataloader = self.replace_sampler(
+                self.train_dataloader = self._prepare_dataloader(
```
Comment (ninginthecloud) [inline pytorch_lightning/trainer/data_loading.py:331]: what if `self.train_dataloader` is not strict Dataloader, but a list of Dataloader? see the following [line351](https://github.com/PyTorchLightning/pytorch-lightning/pull/9721/files#diff-15fd4659acd10d0716002ba2dbfd67aca83dc101d44531c2f234f65140e1e39dR351). 
This is unrelated to this PR, but `reset_train_dataloader` has a lot of operations, the orders of them matter. We need to revisit them to see what is the reasonable order.

**Review by ninginthecloud — COMMENTED** [review]
Thanks for fixing the issue #9655 and refactoring the code, @tchaton ! Overall looks good to me. I have a few minor changes. Additionally, could we also add some docstring for each updated functions, as some function names are not self-explanatory. 
As I'm learning this component, I find there's a lot of logic happening in data_loading.py, I'm curious about how we can sustainably update and maintain this file, as there's a lot of utils functions and they are depending each other, it will be hard to trace. Love to hear your thoughts.

File: pytorch_lightning/trainer/data_loading.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:158]: To improve readability of `prepare_dataloader` which is the exposed utility.

File: pytorch_lightning/trainer/data_loading.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:148]: We didn't have any check. This was done internally. Now it is exposed to the top level for readers to understand when this function is taking effects.

File: pytorch_lightning/trainer/data_loading.py [hunk 5]
```diff
@@ -117,44 +117,69 @@ def auto_add_worker_init_fn(self, dataloader: DataLoader) -> None:
...
+        # the DataLoader should be re-created only if we need to inject
+        # the fault tolerant components or the distributed sampler or we need to track indices for predictions.
         if (
-            self.accelerator_connector.replace_sampler_ddp
-            and self.accelerator_connector.is_distributed
-            and not isinstance(dataloader.sampler, DistributedSampler)
-            and not has_iterable_dataset(dataloader)
+            _fault_tolerant_training()
+            or self._requires_distributed_sampler(dataloader)
+            or mode == RunningStage.PREDICTING
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:150]: Those conditions are used internally with `_get_dataloader_init_kwargs`. It doesn't change the behaviour, just make it apparent within the `prepare_dataloader` when modification would happen.

File: pytorch_lightning/trainer/data_loading.py [hunk 6]
```diff
@@ -267,17 +292,22 @@ def _get_dataloader_init_kwargs(
         return dl_kwargs
 
     @staticmethod
-    def replace_sampler(dataloader: DataLoader, sampler, mode: Optional[RunningStage] = None) -> DataLoader:
+    def _prepare_dataloader(dataloader: DataLoader, sampler, mode: Optional[RunningStage] = None) -> DataLoader:
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:295]: Done !

File: pytorch_lightning/trainer/data_loading.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:170]: `_determine_sampler` doesn't generate the distributed sampler. It resolve the right sampler. Changed it to `_resolve_sampler`.

File: pytorch_lightning/trainer/data_loading.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:331]: Good point. Replace to apply_to_collection.

File: tests/conftest.py [hunk 7]
```diff
@@ -66,6 +66,7 @@ def restore_env_variables():
         "RANK",  # set by DeepSpeed
         "POPLAR_ENGINE_OPTIONS",  # set by IPUPlugin
         # set by XLA
+        "TF2_BEHAVIOR",
```
Comment (carmocca) [inline tests/conftest.py:69]: Same comment as https://github.com/PyTorchLightning/pytorch-lightning/pull/9722#discussion_r717529207

It must be coming from `TensorBoard`, a better solution would be to disable the logger for that test

Comment (tchaton) [reply tests/conftest.py:69]: It seems coming from multiple tests. Let's address this within another PR.

**Review by kaushikb11 — APPROVED** [review]
LGTM!
